### ISSUE TITLE
How can I access an existing SparkContext?

### ISSUE THREAD

**Issue by cacti77** [spark-jobserver/spark-jobserver#167]
I've created a separate streaming job in Spark and I can get hold of the String name of the SparkContext associated with the StreamingContext. Now I want to be able to pass the name of that SparkContext into a Job Server runJob() method so that I can issue SQL queries against it. But how do I get to the actual SparkContext object if all I have is its name?

**Comment by velvia** [comment]
Hi Cacti, hmm....   can you tell me what you mean by "can get hold of the
String name of the SparkContext"?   The problem is that normally the job
server creates the type of context, and does not get it from Spark (and in
fact I'm not sure that's possible, because the StreamingContext is owned by
your job).

The normal pattern would be:  you create a spark streaming job for job
server based on a StreamingContextFactory (see
https://github.com/spark-jobserver/spark-jobserver/pull/142), then you can
have one job that does the streaming, and another job that handles the SQL
queries.   You can then test and develop them separately....

On Wed, Jun 10, 2015 at 5:26 AM, cacti77 <notifications@github.com> wrote:

> I've created a separate streaming job in Spark and I can get hold of the
> String name of the SparkContext associated with the StreamingContext. Now I
> want to be able to pass the name of that SparkContext into a Job Server
> runJob() method so that I can issue SQL queries against it. But how do I
> get to the actual SparkContext object if all I have is its name?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/issues/167>.
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

**Comment by cacti77** [comment]
Hi Velvia. I'm out of the office at the moment so I'll take a look in more detail tomorrow.  I did wonder whether we were going about it the wrong way, so thanks for the tip!  We were planning to use Spark to create the streaming job and then Spark Job Server to query it. So the 2 tasks are still being handled separately.  But I can take a look at using Job Server to create the streaming job as well instead...

Re getting the String name of the SparkContext, for the prototype I was doing, I simply had the Spark app print out the name of the SparkContext it had created. Then I was going to supply that as an argument in my curl request to the jobserver app.  In the longer-term I could have use Job Server's /contexts API call to list the running contexts and find the appropriate one, assuming it had a meaningful name. (Incidentally, is there a Scala equivalent to the /contexts API call so that I can find out the available contexts from directly within my jobserver app?) Either way, though, I still needed some way of getting hold of the actual SparkContext object from its name so that I could create a SQLContext and then run the user's SQL query.

**Comment by velvia** [comment]
Cacti,

 Hm that's an interesting thought.  Exposing the details of the context /
job ID and stuff might be a good feature.  Maybe we should change this
issue into that.

On Wed, Jun 10, 2015 at 10:39 AM, cacti77 <notifications@github.com> wrote:

> Hi Velvia. I'm out of the office at the moment so I'll take a look in more
> detail tomorrow. I did wonder whether we were going about it the wrong way,
> so thanks for the tip! We were planning to use Spark to create the
> streaming job and then Spark Job Server to query it. So the 2 tasks are
> still being handled separately. But I can take a look at using Job Server
> to create the streaming job as well instead...
>
> Re getting the String name of the SparkContext, for the prototype I was
> doing, I simply had the Spark app print out the name of the SparkContext it
> had created. Then I was going to supply that as an argument in my curl
> request to the jobserver app. In the longer-term I could have use Job
> Server's /contexts API call to list the running contexts and find the
> appropriate one, assuming it had a meaningful name. (Incidentally, is there
> a Scala equivalent to the /contexts API call so that I can find out the
> available contexts from directly within my jobserver app?) Either way,
> though, I still needed some way of getting hold of the actual SparkContext
> object from its name so that I could create a SQLContext and then run the
> user's SQL query.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/issues/167#issuecomment-110847645>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

**Comment by cacti77** [comment]
Hi Velvia,
I'm missing something here, and I'm afraid I'm new to Spark and scala:(. In [JobManagerActor.scala](https://github.com/spark-jobserver/spark-jobserver/blob/master/job-server/src/spark.jobserver/JobManagerActor.scala) you already have:

    createContextFromConfig(contextName: String = contextName)

What I would like is a means of getting an existing SparkContext; e.g.,:

    getContextFromConfig(contextName: String = contextName)

...where the returned SparkContext may have been created by another Job Server job or by Spark itself. I don't want to create a new SparkContext or use the one passed into `runJob(sc: SparkContext, config: Config)`.

You must already be listing the SparkContexts when you fulfil requests made to the /contexts REST API call. Please can you tell me where you've implemented that call in your code?

So far I've assumed I need to get hold of the streaming job's SparkContext in order to query it.  If you know of another way to query a streaming job (e.g., via a table name, a named DStream??), I'm open to suggestions.  So far I've only been able to run queries against a stream that was created in the *same* application, whether it be a Spark application or a Job-Server application.  If the query job is isolated from the streaming job I can't see how to pass the SparkContext from the latter to the former.

To me it makes sense to keep the query and streaming jobs separate because one might want to run many different queries against the same streaming job.

**Comment by cacti77** [comment]
Doh! I've been stupid - sorry. I can just pass the name of the existing context in as the value of the `context` parameter on the POST request to the /jobs REST API call. And that context could have been created by either Spark or Job Server originally. (I would prefer it to have been created by Spark for our use case as Job Server will be used as a query engine for a UI layer.) I'll give this a try...

**Comment by cacti77** [comment]
Hmm, this isn't working so well, mainly because it looks like Job Server's /contexts and /jobs REST calls only seem to be able to see contexts and jobs created by Job Server itself! Thus the streaming job created by Spark is simply not visible to Job Server! Is it by design that Job Server can only see contexts or jobs created by itself? Or is that a known limitation which you're intending to address on your roadmap?

Unless this is fixed then I'll be forced to create the streaming job in Job Server rather than Spark.

**Comment by cacti77** [comment]
I'm struggling to get the Job Server streaming and query jobs working, even though I'm now creating a SparkContext in advance via `/contexts` and then passing it into both jobs. The streaming job seems to run ok but the query job fails with "Table Not Found" when I call sql(<my_query>) on the SQLContext object.  In my streaming job I do call `registerTempTable()` on each RDD's DataFrame. In my query app's `runJob()` method I'm doing:
    val sqlContext = new SQLContext(sc)
where `sc` is the argument to `runJob()`. Is this the wrong thing to do?

**Comment by velvia** [comment]
Hi Cacti,

To answer several questions:

1) The job server is not designed to obtain another SparkContext created by
Spark, only the ones created by Job Server itself.  This is more due to the
way Spark is designed -- each Context is its own "application", and the
applications use Spark as a library and are more or less independent.  You
could query a Spark Master for applications, or YARN, or Mesos - but there
is no unified way of getting all of those.

2) As for combining streaming and query jobs I think the reason is that
unless you use Hive, a SQLContext() only remembers the tables registered
for the lifetime of that object.  SInce you have two jobs and the query job
is creating its own SQLContext then it won't have access to tables
registered with the other job.   This is what I would do, is create a
combined SQLStreamingJobContext, something like this:

trait SparkSqlStreamingJob extends SparkJobBase {
  type C = SQLStreamingContext
}

class SQLStreamingContext(c: SparkContext) {
  val streamingContext = new StreamingContext(c, ...)
  val sqlContext = new SQLContext(c)
}

class SQLStreamingContextFactory extends SparkContextFactory {
  import SparkJobUtils._

  type C = SQLStreamingContext with ContextLike

  def makeContext(config: Config, contextConfig: Config, contextName:
String): C = {
    val conf = configToSparkConf(config, contextConfig, contextName)
    new SQLStreamingContext(new SparkContext(conf)) with ContextLike {
      def isValidJob(job: SparkJobBase): Boolean =
job.isInstanceOf[SparkSqlStreamingJob]
      def stop() { this.c.stop() }
    }
  }
}

So what you've done is created a custom SparkContext that has both SQL and
Streaming, and now you create a job that extends SparkSqlStreamingJob, and
you get a SQLStreamingContext where you have access to both the streaming
AND the SQL contexts, so that the state can be shared.

Hope that helps!

On Thu, Jun 11, 2015 at 7:57 AM, cacti77 <notifications@github.com> wrote:

> I'm struggling to get the Job Server streaming and query jobs working,
> even though I'm now creating a SparkContext in advance via /contexts and
> then passing it into both jobs. The streaming job seems to run ok but the
> query job fails with "Table Not Found" when I call sql() on the SQLContext
> object. In my streaming job I do call registerTempTable() on each RDD's
> DataFrame. In my query app's runJob() method I'm doing:
> val sqlContext = new SQLContext(sc)
> where sc is the argument to runJob(). Is this the wrong thing to do?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/issues/167#issuecomment-111163523>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

**Comment by cacti77** [comment]
Hi Velvia,
In response:
1) Thanks for spelling this out.  I hadn't appreciated this important fact until yesterday.  I naively thought that as Job Server provided a REST API for Spark you could use it to access *every* job in the Spark cluster. But I guess it looks like each "application", as you say, is isolated from each other, and that if you want to share stuff between them then you have to persist data to files or tables, etc. However I had noticed https://issues.apache.org/jira/browse/SPARK-6703 was targeted for Spark 1.4.0, and that was released yesterday. So maybe soon it will be possible to share SparkContexts across "applications" after all...?

Now that I understand this limitation (and also eventually remembered you can create the context in advance and pass it into the Job Server jobs directly to share it!), feel free to consider my original question "How can I access an existing SparkContext?" answered now. Thank you.

2) Thanks very much for the code snippets. That's extremely helpful of you:). FYI My context factory actually ended up looking like this:
    class SQLStreamingContextFactory extends SparkContextFactory {
      import SparkJobUtils._
      type C = SQLStreamingContext with ContextLike

      def makeContext(config: Config, contextConfig: Config, contextName: String): C = {
        val batch_interval = config.getInt("batch_interval")
        val conf = configToSparkConf(config, contextConfig, contextName)
        new SQLStreamingContext(new SparkContext(conf), Seconds(batch_interval)) with ContextLike {
          def sparkContext: SparkContext = this.streamingContext.sparkContext
          def isValidJob(job: SparkJobBase): Boolean = job.isInstanceOf[SparkSqlStreamingJob]
          // Stop the streaming context, but not the SparkContext so that it can be re-used
          // to create another streaming context if required:
          def stop() { this.streamingContext.stop(false) }
        }
      }
    }
I'm creating my context like this now:
    curl -d "" "localhost:8090/contexts/stream_sqltest?context-factory=com.my.package.SQLStreamingContextFactory&dependent-jar-uris=file:///path/to/my/stream_sql_2.10-1.0.jar"
but please can you tell me where I should define 'batch_interval'? I keep getting "No configuration setting found for key 'batch_interval'" when I try starting the context. I've tried putting batch_interval in various parts of local.conf (including spark.context-settings), or as a query parameter in the /contexts POST request, but all to no avail so far. Ideally I would like to set it as a parameter on the POST request.  Many thanks for your patience with me on this!

**Comment by velvia** [comment]
Cacti,

Thanks for sharing your mods.  We should contribute this if it's useful to
people.  :)

I would get the batch_interval from contextConfig instead of config.  Then,
you can pass it in the curl command directly, or in the context-settings
section of local.conf etc.   The config is the overall global config, so
you could only set batch_interval at the root level of local.conf.  Hope
that helps!

On Fri, Jun 12, 2015 at 6:35 AM, cacti77 <notifications@github.com> wrote:

> Hi Velvia,
> In response:
> 1) Thanks for spelling this out. I hadn't appreciated this important fact
> until yesterday. I naively thought that as Job Server provided a REST API
> for Spark you could use it to access *every* job in the Spark cluster.
> But I guess it looks like each "application", as you say, is isolated from
> each other, and that if you want to share stuff between them then you have
> to persist data to files or tables, etc. However I had noticed
> https://issues.apache.org/jira/browse/SPARK-6703 was targeted for Spark
> 1.4.0, and that was released yesterday. So maybe soon it will be possible
> to share SparkContexts across "applications" after all...?
>
> Now that I understand this limitation (and also eventually remembered you
> can create the context in advance and pass it into the Job Server jobs
> directly to share it!), feel free to consider my original question "How can
> I access an existing SparkContext?" answered now. Thank you.
>
> 2) Thanks very much for the code snippets. That's extremely helpful of
> you:). FYI My context factory actually ended up looking like this:
> class SQLStreamingContextFactory extends SparkContextFactory {
> import SparkJobUtils._
> type C = SQLStreamingContext with ContextLike
>
>   def makeContext(config: Config, contextConfig: Config, contextName: String): C = {
>     val batch_interval = config.getInt("batch_interval")
>     val conf = configToSparkConf(config, contextConfig, contextName)
>     new SQLStreamingContext(new SparkContext(conf), Seconds(batch_interval)) with ContextLike {
>       def sparkContext: SparkContext = this.streamingContext.sparkContext
>       def isValidJob(job: SparkJobBase): Boolean = job.isInstanceOf[SparkSqlStreamingJob]
>       // Stop the streaming context, but not the SparkContext so that it can be re-used
>       // to create another streaming context if required:
>       def stop() { this.streamingContext.stop(false) }
>     }
>   }
> }
>
> I'm creating my context like this now:
> curl -d ""
> "localhost:8090/contexts/stream_sqltest?context-factory=com.my.package.SQLStreamingContextFactory&dependent-jar-uris=file:///path/to/my/stream_sql_2.10-1.0.jar"
> but please can you tell me where I should define 'batch_interval'? I keep
> getting "No configuration setting found for key 'batch_interval'" when I
> try starting the context. I've tried putting batch_interval in various
> parts of local.conf (including spark.context-settings), or as a query
> parameter in the /contexts POST request, but all to no avail so far.
> Ideally I would like to set it as a parameter on the POST request. Many
> thanks for your patience with me on this!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/issues/167#issuecomment-111494637>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

**Comment by cacti77** [comment]
Hi Velvia,
Thanks for the info.  Yes, I can confirm that using:

    val batch_interval = contextConfig.getInt("batch_interval")
in SQLStreamingContextFactory and then doing:

    curl -d "" "localhost:8090/contexts/stream_sql_test?context-factory=com.my.package.SQLStreamingContextFactory&dependent-jar-uris=file:///path/to/my/stream_sql_2.10-1.0.jar&batch_interval=10"
actually worked! I was able to start the streaming job in the context 'stream_sql_test' and then start dropping text files into a watched folder. If I kept repeating the curl command that issued the SQL query against the same context I could see it returning results against the new files.

Thanks very much for your help getting this running. Yes, feel free to turn this into a sample if you wish. Omitting packing names, I changed SQLStreamingContext.scala to:

    import com.typesafe.config.Config
    import org.apache.spark.SparkContext
    import org.apache.spark.sql.SQLContext
    import org.apache.spark.streaming.{Seconds, StreamingContext}
    import spark.jobserver.{ContextLike, SparkJobBase}
    import spark.jobserver.context.SparkContextFactory
    import spark.jobserver.util.SparkJobUtils

    class SQLStreamingContextFactory extends SparkContextFactory {
      import SparkJobUtils._
      type C = SQLStreamingContext with ContextLike
      def makeContext(config: Config, contextConfig: Config, contextName: String): C = {
        val batch_interval = contextConfig.getInt("batch_interval")
        val conf = configToSparkConf(config, contextConfig, contextName)
        new SQLStreamingContext(new SparkContext(conf), Seconds(batch_interval)) with ContextLike {
          def sparkContext: SparkContext = this.streamingContext.sparkContext
          def isValidJob(job: SparkJobBase): Boolean = job.isInstanceOf[SparkSqlStreamingJob]
          // Stop the streaming context. See also
          // https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.streaming.StreamingContext
          // regarding whether or not the associated SparkContext will also be stopped.
          def stop() { this.streamingContext.stop() }
        }
      }
    }

I didn't change your SparkSqlStreamingJob at all:

    import spark.jobserver.SparkJobBase
    trait SparkSqlStreamingJob extends SparkJobBase {
      type C = SQLStreamingContext
    }

but I did add an extra constructor parameter to your SQLStreamingContext for the batch interval:

    import org.apache.spark.SparkContext
    import org.apache.spark.sql.SQLContext
    import org.apache.spark.streaming.{Duration, StreamingContext}

    class SQLStreamingContext(sc: SparkContext, batchInterval: Duration) {
      val streamingContext = new StreamingContext(sc, batchInterval)
      val sqlContext = new SQLContext(sc)
    }

This is the streaming job:

    import com.typesafe.config.Config
    import org.apache.spark.SparkContext
    import org.apache.spark.rdd.RDD
    import org.apache.spark.sql.{SQLContext, Row}
    import spark.jobserver.{SparkJob, SparkJobValidation, SparkJobValid, SparkJobInvalid}
    import org.apache.spark.streaming.{Minutes, Seconds, StreamingContext, Time}
    import scala.util.Try

    object StreamFilesIntoTableJob extends SparkSqlStreamingJob {

      override def validate(sqlsc: SQLStreamingContext, config: Config): SparkJobValidation = {
        Try(config.getString("directory"))
          .map(x => SparkJobValid)
          .getOrElse(SparkJobInvalid("Missing 'directory' configuration parameter!"))
      }

      override def runJob(sqlsc: SQLStreamingContext, config: Config): Any = {
        // Debugging (look in Spark Job Server's console output):
        val directory = config.getString("directory")
        println("directory  = " + directory)
        
        // Create a stream on the specified directory.
        // Note that any files already in the directory will be ignored!
        // Only new files that arrive during a given batch interval count!
        val lines = sqlsc.streamingContext.textFileStream(directory)
        val data = lines.map(_.split(","))

        // Convert RDDs of the data DStream to DataFrame and run SQL query
        data.foreachRDD((rdd: RDD[Array[String]], time: Time) => {
          // Get the singleton instance of SQLContext
          val sqlContext = sqlsc.sqlContext
          // This is used to implicitly convert an RDD to a DataFrame:
          import sqlContext.implicits._
          // Convert RDD[Array[String]] to RDD[Person] to DataFrame
          val people = rdd.map(p => Person(p(0), p(1).trim.toInt)).toDF()
          // Register as table
          people.registerTempTable("people")    
        })
        sqlsc.streamingContext.start()
        sqlsc.streamingContext.awaitTermination()
      }
    }

    // Use a case class to define the schema for the data being read in from the files.
    case class Person(name: String, age: Int)

This is the query job:

    import com.typesafe.config.{Config, ConfigFactory}
    import org.apache.spark.{SparkContext}
    import org.apache.spark.sql.{SQLContext, Row}
    import spark.jobserver.{SparkJob, SparkJobValidation, SparkJobValid, SparkJobInvalid}
    import scala.util.Try

    object QueryStreamJob extends SparkSqlStreamingJob {

      override def validate(sqlsc: SQLStreamingContext, config: Config): SparkJobValidation = {
        Try(config.getString("sql_query"))
          .map(x => SparkJobValid)
          .getOrElse(SparkJobInvalid("Missing sql_query configuration parameter!"))
      }

      override def runJob(sqlsc: SQLStreamingContext, config: Config): Any = {
        // Debugging (look in Spark Job Server's console output):
        val query = config.getString("sql_query")
        println("sql_query  = " + query)
        
        val results = sqlsc.sqlContext.sql(query)
        // Return the first 10 rows:
        results.head(10)
      }
    }

Sample text file, people1.txt:

    Michael, 29
    Andy, 30
    Justin, 19
    Fred, 15

Sample SQL query for the POST request to the query job:

    sql_query="SELECT * FROM people WHERE age >= 13 AND age <= 19"

Sample results from a query run soon after people1.txt was dropped into the watched folder:

    curl --data-binary @query_post_dat.txt "localhost:8090/jobs?context=stream_sql_test&appName=query_stream&classPath=com.ge.ogaw.QueryStreamJob&sync=true"
    {
      "status": "OK",
      "result": ["[Justin,19]", "[Fred,15]"]
    }

I hope this is useful to you and others.

**Comment by cacti77** [comment]
The only caveats on the previous comment is that:

1) Normally I run Job Server from a docker container. Unfortunately the curl request to create the context fails with ClassNotFoundException for SQLStreamingContextFactory when I run Job Server from docker. But it works fine if I run Job Server normally, outside of a docker container. Maybe it's the way dependent-jar-uris is interpreted by docker. I'll take a look at that...

2) When I'm done I thought that killing the context with:

   curl -X DELETE localhost:8090/contexts/stream_sql_test
would kill the context and all the jobs in it. It does kill the context but the streaming job happily stays in the RUNNING state! If I try to kill it explicitly with:

   curl -X DELETE localhost:8090/jobs/<job_ID>
I get "The server was not able to produce a timely response to your request."  I can only seem to kill the job by restarting Job Server. Are there any known issues with killing contexts?

**Comment by velvia** [comment]
@cacti77 I haven't played much with streaming, I wonder if it is because something is not shut down properly.  In the code I see that the sparkContext and sqlContext themselves are not stopped.

**Comment by cacti77** [comment]
@velvia FYI I managed to get job-server working inside docker with file streaming as well but there were a couple gotchas to get past. First I had to use the `docker run -v` option to mount a host directory; put the jar file in there, and then use the container's internal path as the value for the `dependent-jar-uris` parameter. That got past the ClassNotFoundException for SQLStreamingContextFactory.

But the second, harder, gotcha was that the watched directory into which I dropped the data files had to be visible to both Job Server running inside docker *and* to the Spark executors running outside docker.

Re stopping the context I can confirm that the SparkContext *is* stopped in Job Server when I issue the delete request against the context:

    [2015-06-16 14:14:47,921] INFO  .spark.storage.MemoryStore [] [] - MemoryStore cleared
    [2015-06-16 14:14:47,923] INFO  spark.storage.BlockManager [] [] - BlockManager stopped
    [2015-06-16 14:14:47,925] INFO  storage.BlockManagerMaster [] [] - BlockManagerMaster stopped
    [2015-06-16 14:14:47,929] INFO  .apache.spark.SparkContext [] [] - Successfully stopped SparkContext

(That the SparkContext is stopped is consistent with the Scaladoc for StreamingContext.stop().)

Even though the corresponding StreamFilesIntoTableJob isn't stopped in Job Server that doesn't stop me doing anything so I leave it up to you whether you want to investigate that further. Many thanks again for all your fantastic help!

**Comment by bistaumanga** [comment]
Hi, This discussion was really helpful for creating something like we needed.

However, we want to to run sql job forever like streaming job once submitted. Can we make some adjustments for this? As i see, we cannot add anything to streaming-context once started, right?

**Comment by cacti77** [comment]
@bistaumanga I'm not sure I understand what you want to do and it's been a while since I looked into this. I'd just encourage you to have a play and try out your ideas for yourself:).

**Comment by rupeshkarna** [comment]
Thanks for your code snippet. It was very useful to us.

Here, the streaming job is running forever but when we query using the sql job, it returns the result only once. Is there any way so that we can run the sql job forever just like the streaming job and get the results of desired query continuously?

**Comment by cacti77** [comment]
OK, now I get it:). I haven't tried this myself and I'm pretty rusty on Job Server stuff as I haven't used it for a while. But the main approaches I can think of off the top of my head are:

1. Simply run the query inside the streaming job itself (right after `people.registerTempTable("people")`. If you use println() to print out the query results you could then simply follow Job Server's logs (e.g., with `tail -f`) and see the query results every batch interval. Or you could write the query results to a database or file in a folder and run a database trigger or folder watchers to take action when new data or a file appears.

2. Use a `cron` job or some other scheduler to run the existing query job for you at specified times or intervals.

3. Put a loop inside the query job, getting it to write the results out to a database or file every few seconds. Then invoke the query job asynchronously.

**Comment by velvia** [comment]
We could potentially provide a template streaming job that accepts a function that can
convert the incoming dstream => stream of RDDs => DataFrames

then accept a SQL function to query that DataFrame continuously.

**Comment by bistaumanga** [comment]
The use case @rupeshkarna is talking is like this:

We have continuous stream of data in kafka, we want to run some aggregations on dstream and write the results back to kafka. We could do this by putting all aggregations together in a same spark-streaming jobs or dividing them among few streaming jobs.

But we want different users to add their own jobs dynamically. Spark job server allows this for batch mode. But, i couldn't find a way to share a DStream between users (like NamedRDD).

So as per this discussion, we could ETL data from kafka to Inmemory Spark Dataframes and perform sql query on those job. But we need the jobs to be running continuously forever until users kill them.
 
Scheduling with cronjob/chronos is one solution. But we are not sure about it's latency implcations. Maybe something inbuild would be better. And we want this for many jobs, 500-1000 aggregations without compromising the latency, 5-15 secs batching.

Maybe kafka+Spark combination is not fit for this usecase. Any Suggestions?

**Comment by velvia** [comment]
Hmmmm….. no, I think sharing the dstream is likely what you want.

Could you achieve this via a custom StreamingContext?
Have a look at StreamingContextFactory, but instead of returning a StreamingContext, return a MyCustomStreamingContext which defines an additional method, for example,

getMyDstream(…)
setMyDstream(…)

Then your jobs could access this method to get at the shared stream.

-Evan

**Comment by usufarif** [comment]
I tried several options, this is what worked eventually for me..

object SomeName extends App {

val conf = new SparkConf()...
val sc = new SparkContext(conf)

implicit val sqlC = SQLContext.getOrCreate(sc)
getDF1(sqlC)

def getDF1(sqlCo: SQLContext): Unit = {
    val query1 =  SomeQuery here  
    val df1 = sqlCo.read.format("jdbc").options(Map("url" -> dbUrl,"dbtable" -> query1)).load.cache()

 //iterate through df1 and retrieve the 2nd DataFrame based on some values in the Row of the first DataFrame

  df1.foreach(x => {
    getDF2(x.getString(0), x.getDecimal(1).toString, x.getDecimal(3).doubleValue) (sqlCo)
  })     
}

def getDF2(a: String, b: String, c: Double)(implicit sqlCont: SQLContext) :  Unit = {
  val query2 = Somequery

  val sqlcc = SQLContext.getOrCreate(sc)
  //val sqlcc = sqlCont //Did not work for me. Also, omitting (implicit sqlCont: SQLContext) altogether did not work
  val df2 = sqlcc.read.format("jdbc").options(Map("url" -> dbURL, "dbtable" -> query2)).load().cache()
   .
   .
   .
 }
}

Note: In the above code, if I omitted (implicit sqlCont: SQLContext) parameter from getDF2 method signature, it would not work. I tried several other options of passing the sqlContext from one method to the other, it always gave me NullPointerException or Task not serializable Excpetion. Good thins is it eventually worked this way, and I could retrieve parameters from a row of the DataFrame1 and use those values in loading the DataFrame 2.
